src/EFCore.PG/Utilities/SortOrderHelper.cs

```diff
@@ -0,0 +1,48 @@
+using System.Linq;
+using JetBrains.Annotations;
+using Npgsql.EntityFrameworkCore.PostgreSQL.Metadata;
+
+namespace Npgsql.EntityFrameworkCore.PostgreSQL.Utilities
+{
+    static class SortOrderHelper
+    {
+        public static bool IsDefaultSortOrder([NotNull] SortOrder[] sortOrders)
```

Comment:
Looking at this again, it seems like we're encoding the assumption that the default sort order is ascending - this method is wrong if there's ever a case where it's not. If we're really OK with making that assumption, then as I wrote above I'm not sure why we need the Unspecified - we can just not generate anything in the migration for ascending, for cleaner SQL. If we really don't trust that assumption, then we shouldn't omit the annotation either when all columns are descending.

I think that unless we have a reason to think otherwise, the assumption does seem to make sense (and be based on the PostgreSQL docs). Some research may help confirm this, am not sure.

I know I'm repeating what I wrote above :) I do think we should decide if we assume or not, and go all the way. 

For null sort order we do need Unspecified (or rather Default) as a way of saying "the default for whatever the sort order says". When we scaffold, we can check the sort order and scaffold Default rather than the specific NullsFirst/Last.

Reply:
> this method is wrong if there's ever a case where it's not.

Yes, and there are certainly scenarios where that's the case. I.e. any index access method that doesn't support ordering (anything except `btree` of the built-in AMs). Postgres will error if you try to use ASC/DESC with such index AMs though. I'm not sure how we can prevent that. 

The Postgres documentation is pretty clear about the default values and the requirements for index AMs that support ordering. So I think this is a safe assumption to make. The only thing we could do to harden this even more is to somehow query the DB for index AMs and detect whether the selected `Method` supports ordering before generating migration SQL.

Should I just go for removing `Unspecified` and not emit any SQL for ASC?

Reply:
At that point, it's hard to argue for using an enum as well (two values).

Reply:
I can see two arguments for keeping `Unspecified`:

1. Consistency between the enums.
2. It's likely that EF Core will have such value, given SqlClient does, and we'd likely adopt theirs at some point.

Reply:
OK, thanks for doing the research on this!

> The only thing we could do to harden this even more is to somehow query the DB for index AMs and detect whether the selected `Method` supports ordering before generating migration SQL.

While we can easily query this when scaffolding, I don't think we can do that for migrations. In EF Core, migration generation happens without any database connection, so there's no way to actually do this (consider that a migration SQL script can be generated without even having a database nearby).

This means that we need a hardcoded table of which index access methods support ordering and which don't. I don't think it's an actual problem, as this is very unlikely to change, and if it does, we can always update this table upon request. And if we need this for migration generation, we may as well also use it when scaffolding, even if we *do* have a database connection during the latter.

> Should I just go for removing `Unspecified` and not emit any SQL for ASC?

I think that makes sense. About your proposed reasons for keeping the enum:

> Consistency between the enums.

I don't really think there's any reason these two enums have to be consistent... SortOrder simply has a default value, whereas NullSortOrder's default depends on the other enum - that's a good enough justification for me, unless you feel strongly about this.

> It's likely that EF Core will have such value, given SqlClient does, and we'd likely adopt theirs at some point.

If we switch to an EF Core enum, then in any case we'll have a breaking change on our hands and can do that change at that point. I'm still hoping this can be done on the EF Core side for 3.0, so that we can align - any chance you're interested in submitting that?

> At that point, it's hard to argue for using an enum as well (two values).

I'm not sure - have ascending/descending still seems better than true/false, where one can get confused on which direction goes where etc. I'd keep the enum, but generate nothing in SQL for ascending. And as already done, if all values in the annotation are ascending, the annotation itself can be omitted.

Reply:
> any chance you're interested in submitting that?

I started looking at it, but had trouble building the sln. I'll look into it more.

Reply:
Ok, be sure to run the restore script at the door first. After that, to start visual studio use the starts.cmd script with the runtime solution. If you run into more issues let me know.

Reply:
> This means that we need a hardcoded table of which index access methods support ordering and which don't.

Where would you put this table. I'm thinking it would be nice if devs could modify this list to unblock themselves in case they encounter an access method that supports sorting, but we don't include.

Reply:
I really don't think that's an actual problem we need to spend time on... I've been working on PostgreSQL support and I don't remember when a new index access method was last introduced... We can hardcode this for now, and it's always possible to expose it later if we need to - although I suspect it will be easier for users too if we just update that table if/when it needs updating...

Reply:
OK, so in reality, I can just do a `string.Equals(method, "btree")` and be done with it?

Reply:
Yeah, I'd say that's the way to go... We can always modify this later if we need to - although I suspect this won't happen...